onmt/translate/translator.py

```diff
@@ -277,6 +280,143 @@ def translate(
+        self.model.decoder.map_state(
+            lambda state, dim: tile(state, beam_size, dim=dim))
+        if isinstance(memory_bank, tuple):
+            memory_bank = tuple(tile(x, beam_size, dim=1) for x in memory_bank)
+            mb_device = memory_bank[0].device
+        else:
+            memory_bank = tile(memory_bank, beam_size, dim=1)
+            mb_device = memory_bank.device
+
+        memory_lengths = tile(src_lengths, beam_size)
+        src_map = (tile(batch.src_map, beam_size, dim=1)
+                   if use_src_map else None)
+
+        seq_so_far = torch.full([batch_size, 1], start_token, dtype=torch.long,
+            device=mb_device)
+        alive_attn = None
+
+        # Give full probability to the first beam on the first step.
+        topk_log_probs = torch.tensor([0.0], device=mb_device).repeat(batch_size)
+
+        # Structure that holds finished sequence for each batch index.
+        hypotheses = [[] for _ in range(batch_size)]
+
+        for step in range(max_length):
+            decoder_input = seq_so_far[:, -1].view(1, -1, 1)
+
+            log_probs, attn = self._decode_and_generate(
+                decoder_input,
+                memory_bank,
+                batch,
+                data,
+                memory_lengths=memory_lengths,
+                src_map=src_map,
+                step=step,
+                batch_offset=torch.arange(batch_size, dtype=torch.long)
+            )
+
+            vocab_size = log_probs.size(-1)
+
+            if step < min_length:
+                log_probs[:, end_token] = -1e20
+
+            actual_log_probs = log_probs - (log_probs - F.log_softmax(log_probs)).mean(1, True).repeat([1, log_probs.shape[1]])
+
+            dist = torch.distributions.Multinomial(logits=log_probs, total_count=1)
+            topk_ids = torch.argmax(dist.sample(), dim=1, keepdim=True)
+            topk_scores = log_probs.gather(dim=1, index=topk_ids) 
+
+            # Recover log probs.
+            topk_log_probs = topk_scores
+
+            # Append last prediction.
+            seq_so_far = torch.cat(
+                [seq_so_far,
+                topk_ids.view(-1, 1)], -1)
+            if return_attention:
+                current_attn = attn
+                if alive_attn is None:
+                    alive_attn = current_attn
+                else:
+                    alive_attn = torch.cat([alive_attn, current_attn], 0)
+
+        is_finished = topk_ids.eq(end_token)
+        if step + 1 == max_length:
+            is_finished.fill_(1)
+
+        # Save finished hypotheses.
+        is_finished = is_finished.to('cpu')
+
+        predictions = seq_so_far.view(-1, 1, seq_so_far.size(-1))
+        attention = (
+            alive_attn.view(
+                alive_attn.size(0), -1, 1, alive_attn.size(-1))
+            if alive_attn is not None else None)
+
+        for i in range(is_finished.size(0)):
+            # Store finished hypotheses for this batch.
```

Comment:
@daphnei  Why aren't you checking is_finished[i] here before appending to results? In original code, we also remove the finished sentences from batch for efficiency . Why isn't that done?

Reply:
This is a valid comment, but this can be added later if needed. I expect that removing finished batches is less important in non beam search decoding.

Reply:
I was mostly trying to simplify the code. Unless, the batch sizes is really large, I didn't think it would make much of a performance difference as there is only ever one "path" per example.